<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
 <HEAD>
   <TITLE> [c++-pthreads] Re: thread-safety definition
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:c%2B%2B-pthreads%40codesourcery.com?Subject=Re%3A%20%5Bc%2B%2B-pthreads%5D%20Re%3A%20thread-safety%20definition&In-Reply-To=%3C20040108162036.GG23908%40diablo.name%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <style type="text/css">
       pre {
           white-space: pre-wrap;       /* css-2.1, curent FF, Opera, Safari */
           }
   </style>
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="000184.html">
   <LINK REL="Next"  HREF="000169.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[c++-pthreads] Re: thread-safety definition</H1>
    <B>Ted Baker</B> 
    <A HREF="mailto:c%2B%2B-pthreads%40codesourcery.com?Subject=Re%3A%20%5Bc%2B%2B-pthreads%5D%20Re%3A%20thread-safety%20definition&In-Reply-To=%3C20040108162036.GG23908%40diablo.name%3E"
       TITLE="[c++-pthreads] Re: thread-safety definition">baker at cs.fsu.edu
       </A><BR>
    <I>Thu Jan  8 16:20:36 UTC 2004</I>
    <P><UL>
        <LI>Previous message: <A HREF="000184.html">[c++-pthreads] Re: thread-safety definition
</A></li>
        <LI>Next message: <A HREF="000169.html">thread-safety definition
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#163">[ date ]</a>
              <a href="thread.html#163">[ thread ]</a>
              <a href="subject.html#163">[ subject ]</a>
              <a href="author.html#163">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>&gt;<i> A catch(...) without a re-throw isn't necessarily a silly thing to do, 
</I>&gt;<i> especially in a destructor.  Sometimes, the only alternative is to risk 
</I>&gt;<i> a call to terminate().
</I>
For this reason, Ada requires that abort (what we are calling
cancellation here) be deferred automatically during the execution
of finalizers (what we are calling destructors here).

It is important that, as you say, finalizers can catch exceptions
and encapsulate them, allowing finalization to continue with the
next finalizer.

This is an important point, that one may have recursive/nested
exceptions, i.e., while handling one exception another exception
may be raised and handled, upon completion of which processing of
the the first exception resumes.

In particular, if one is executing a chain of
finalizers/descructors due to exception propagation causing one to
exit the scope of an object with finalization, the
finalizer/destructor should be able to catch local exceptions that
might be raised/thrown by its own code, without re-raising them.
On return from the finalizer/destructor the propagation of the
other exception then continues, unaffected.

This is what you also want with abort/cancellation.  In the
processing of the cancellation you execute cleanup handlers,
finalizers, desctructors (or whatever you choose to call them)
as you unwind the stack.  Each of those may generate new
exceptions, and handle them locally.  That does not stop the
unwinding of the stack until the original abort/cancellation
is caught and handled (if you allow that to be caught and
handled) or the thread terminates.

--Ted


</PRE>

<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="000184.html">[c++-pthreads] Re: thread-safety definition
</A></li>
	<LI>Next message: <A HREF="000169.html">thread-safety definition
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#163">[ date ]</a>
              <a href="thread.html#163">[ thread ]</a>
              <a href="subject.html#163">[ subject ]</a>
              <a href="author.html#163">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://sourcerytools.com/cgi-bin/mailman/listinfo/c++-pthreads">More information about the c++-pthreads
mailing list</a><br>
</body></html>
